usr/lib/cca_stdll/cca_specific.c

```diff
@@ -3389,6 +3513,10 @@ CK_RV token_specific_init(STDLL_TokData_t * tokdata, CK_SLOT_ID SlotNumber,
     if (rc != CKR_OK)
         goto error;
 
+    rc = cca_get_min_card_level(tokdata);
```

Comment:
Call this in cca_handle_apqn_event() as well. 

But then it needs proper locking:

- Add `pthread_rwlock_t min_card_version_rwlock;` to `struct cca_private_data`
- Initialize the lock in `token_specific_init()`: `pthread_rwlock_init(&cca_private->min_card_version_rwlock, NULL);` (plus error checking)
- Free lock in token_specific_final(): `pthread_rwlock_destroy(&cca_private->min_card_version_rwlock);`
- Write-Lock in cca_get_min_card_level() where you copy the version from `card_level_data` to `cca_private->min_card_version`: `pthread_rwlock_wrlock(&cca_private->min_card_version_rwlock);` plus error checking and `pthread_rwlock_unlock(&cca_private->min_card_version_rwlock);` after the copy.  Do not hld the lock while you iterate over the APQNs, just while you are copying the version from `card_level_data` to `cca_private->min_card_version`.
- Read lock in `ccatok_pkey_get_firmware_wkvp() `where you chekc the min-card_version: `pthread_rwlock_rdlock(&cca_private->min_card_version_rwlock);` plus error checking and `pthread_rwlock_unlock(&cca_private->min_card_version_rwlock);`. I.e. its this line that needs the lock:  `if (cca_data->min_card_version.ver < 7) {`
- in `cca_handle_apqn_event()` call `cca_get_min_card_level()` as well.
- in `cca_handle_apqn_event()` check resulting min_card_version and determine the right value of `cca_data->pkey_wrap_supported` again. I.e. if min_card_version < 7, then set pkey_wrap_supported = FALSE. If min_card_version >= 7, then check msa level, etc again and set pkey_wrap_supported = TRUE if all preconditions are meet (no need to get FW-WKVP again here).

Reply:
Hm, if cca_handle_apqn_event() detects that pkey is now possible, but it has never been possible before, then the FW-WKVP is not available. So you would need to to ccatok_pkey_get_firmware_wkvp() as well. 
But then the FW-WKVP field needs to be lock-protected as well against multithreaded access.
Same for the pkey_wrap_supported field actually....

Reply:
See PR https://github.com/opencryptoki/opencryptoki/pull/677 for how this can be done and how pkey_wrap_supported can be set without additional locks in the APQN event routine.